default/scripting/buildings/TERRAFORM.focs.txt

```diff
@@ -84,5 +84,80 @@ TARGET_TERRAFORMING
     icon = "icons/building/terraform.png"
 '''
 
+BuildingType
+    name = "BLD_TERRAFORM_BEST"
+    description = "BLD_TERRAFORM_BEST_DESC"
+    buildcost =  75 * (1 - 0.4 * Statistic If condition = And [Source EmpireHasAdoptedPolicy empire = Source.Owner name = "PLC_TERRAFORMING"])
+                    * Target.HabitableSize
+                      // calculating multiple steps: 1+2+..+N == (1+N)*N/2
+                    * max(1, (1 + (PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType))
+                             * (PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType) / 2 )
+                    * [[BUILDING_COST_MULTIPLIER]]
+                    * (1 + (2 * Statistic If condition = And [Source EmpireHasAdoptedPolicy empire = Source.Owner name = "PLC_ENVIRONMENTALISM"]))
+    buildtime =  (8 - (4 * Statistic If condition = And [Source EmpireHasAdoptedPolicy empire = Source.Owner name = "PLC_TERRAFORMING"]))
+               * PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType
+    location = And [
+        Planet
+        Not Contains Building name = "BLD_TERRAFORM"
+[[FOR_ALL_TERRAFORM_PLANET_TYPES(        Not Contains Building name = "BLD_TERRAFORM_,")]]
+[[FOR_ALL_TERRAFORM_PLANET_TYPES(        Not Enqueued type = Building name = "BLD_TERRAFORM_,")]]
+        OwnedBy empire = Source.Owner
+        TargetPopulation low = 1
+        Not Planet type = [Asteroids GasGiant]
+        Not Planet environment = [Uninhabitable Good]
+        Not Contains Building name = "BLD_GAIA_TRANS"
+        Not HasSpecial name = "GAIA_SPECIAL"
+        // Hide building if the species can not improve (Check species' highest available target environment tag)
+        Not HasTag name = "NO_TERRAFORM"
+        Not And [
+            HasTag name = "TERRAFORM_TO_POOR"
+            Or [ Planet environment = Adequate Planet environment = Poor ]
+        ]
+        Not And [
+            HasTag name = "TERRAFORM_TO_ADEQUATE"
+            Or [ Planet environment = Adequate ]
```

Comment:
Can't you just specify something like:
(PlanetTypeDifference from = LocalCandidate.PlanetType to = LocalCandidate.NextBestPlanetType) > 0

Reply:
sounds better